File: SI.py
import numpy as np
import json
import pickle
import sys
from scipy.optimize import curve_fit


class SI:
    def __init__(self, **conf):

        # reading params from the config conf, otherwise set to default
        # parameter 'p' of the Bass model
        self.p = conf.get("p", 0.000104)
        # parameter 'q' of the Bass model
        self.q = conf.get("q", 0.12)
        # number of timesteps
        self.stop = conf.get("stop", 129)
        # correcting the threshold
        self.high = conf.get("high", 0)
        self.low = conf.get("low", 0)
        # total number of runs with the above parameter settings
        self.num_runs = conf.get("num_runs", 1)

        # reading in byte-saved variables created by preferences.py that are the same for each run
        with open('adjacency.pickle', 'rb') as f:
            # A is the adjacency matrix
            # indexmap maps user_ids to the row indices of A
            # indexmap_back maps row indices of A to user_ids
            # seed is a boolean vector of the initial infected nodes
            self.A, self.indexmap, self.indexmap_back, self.seed = pickle.load(f)

        # creating the necessary variables, setting the initial infected nodes
        self.restart()

    def restart(self):
        # init new simulation
        self.time_counter=1

        # creating infected and susceptible lists
        self.infected = np.zeros(self.A.shape[0], dtype=bool)
        # initial infection
        self.infected[self.seed] = True
        # initial susceptibles, aka everybody who's not infected
        self.susceptible = ~self.infected

        # storing the infection time for the nodes, setting infected nodes times to 1
        self.time_infected = np.array(self.infected, dtype=int)

        # storing number of infected neighbors for nodes
        self.node_neighborhood_num_infected = np.matrix(np.zeros_like(self.infected, dtype=float))
        # storing total number of neighbors
        self.node_neighborhood_num_neighbors = self.A.sum(axis=0)
        # adding 1 for isolated nodes to avoid division by 0
        self.node_neighborhood_num_neighbors[(self.node_neighborhood_num_neighbors == 0)[0]] = 1
        # counting number of infected neighbors for initial infection
        self.node_neighborhood_num_infected += self.A[self.infected].sum(axis=0)

    def transform(self, a):
        """
        Function that is able to correct for neighborhood infection rate.
        """

        x = [0, 0.5, 1]
        y = [1-self.low, 1+self.high, 1-self.low]
        params = np.polyfit(x, y, 2)

        return np.polyval(params, a)*np.array(a)

    def step_time(self):
        # increment time counter
        self.time_counter += 1

        # fraction of infected neighbors vs total neighbors
        a = self.node_neighborhood_num_infected/self.node_neighborhood_num_neighbors
        # uniform random number between 0 and 1
        r = np.random.rand(a.shape[1])
        # calculating threshold, comparing random number to threshold
        mask = (r < self.p + self.q*self.transform(a))
        # infect nodes where random number is smaller than the threshold and that are susceptible
        new_infected = np.array(self.susceptible & mask)[0]

        # store infection time for newly infected nodes
        self.time_infected[new_infected] = self.time_counter
        # union of old and new infections
        self.infected = self.infected | new_infected
        # decreasing the susceptibles with the newly infected nodes
        self.susceptible = self.susceptible & (~new_infected)
        # incrementing the counters in the array storing the number of infected neighbors
        self.node_neighborhood_num_infected += self.A[new_infected].sum(axis=0)

        return new_infected.sum()

    def run_new(self):
        # init simulation, time_counter = 1
        self.restart()
        # stepping time
        for t in range(2, self.stop):
            n = self.step_time()

        # storing results, raw output
        return {
            "time_infected": {str(self.indexmap_back[i]): int(t) for i, t in enumerate(self.time_infected) if t!=0},
            "p": self.p,
            "q": self.q,
            "stop": self.stop,
            "num_nodes": self.A.shape[0],
            "high": self.high,
            "low": self.low
        }

    def run_batch(self):
        # storing results of multiple runs in a dictionary
        o = {}
        # running the simulation multiple times, same parametrization
        for run in range(self.num_runs):
            res = self.run_new()
            o["run_" + str(run).zfill(2)] = res

        return o

    def average_batch(self, o):
        timelines = []

        for r in o:
            node_adoptions = o[r]["time_infected"]

            # getting adoption time histogram
            timeline = np.zeros(self.stop)
            for k in node_adoptions.keys():
                timeline[node_adoptions[k] - 1] += 1
            timelines.append(timeline)

        return {
            "avg_timeline": list(np.array(timelines).mean(axis=0)),
            "p": self.p,
            "q": self.q,
            "stop": self.stop,
            "num_nodes": self.A.shape[0],
            "high": self.high,
            "low": self.low
        }


class DE_fit:
    def __init__(self, avg_output):
        self.t = np.array(range(1,avg_output["stop"]+1))
        self.timeline = avg_output["avg_timeline"]
        self.cum_timeline = np.cumsum(self.timeline)
        self.N = avg_output["num_nodes"]
        self.p = avg_output["p"]
        self.q = avg_output["q"]

    def bass_solution(self, t, P, Q):
        # Bass DE solution
        return self.N * (1-np.exp(-(P+Q)*t))/(1+Q/P*np.exp(-(P+Q)*t))

    def fit(self):
        param, param_cov = curve_fit(self.bass_solution, self.t, self.cum_timeline, p0=[self.p, self.q])
        return param, param_cov

if __name__ == "__main__":
    # reading config from standard input, if None, then values are default
    conf = json.load(sys.stdin)
    # initializing simulation with values from conf
    simulation = SI(**conf)

    # storing results of multiple runs in a dictionary
    output = {}
    # running the simulation multiple times, same parametrization
    for r in range(simulation.num_runs):
        result = simulation.run_new()
        output["run_"+str(r).zfill(2)] = result

    print(json.dumps(output))

File: R_version/SI.R
library(data.table)
library(Matrix)
library(tictoc)

mode <- "iwiw"

init <- function(){
# This is the function, where we can set up the global parameters of the whole simulation.

    # parameter 'p' of the Bass model
    p_par <<- 0.000104
    # parameter 'q' of the Bass model
    q_par <<- 0.12
    # number of timesteps
    stop_par <<- 128

    # total number of runs with the above parameter settings
    num_runs <<- 1

    # creating necessary tables
    if (mode == "karate"){
        # this is the mode where we read the Zachary network edgelist for educational/debugging purposes
        # number of nodes
        N <<- 34
        elist <<- fread("R_karate.csv")
        # the edgelist only lists edges once, to create a full adj matrix, we have to give the reverse column
        # order as well to the sparse matrix constructor
        A <<- sparseMatrix(c(elist$V1,elist$V2),c(elist$V2,elist$V1),dims=c(N,N))
        # initial infected nodes
        seed <<- as.vector(c(1,2))
    }
    
    # R indexes from 1!!!
    if (mode == "iwiw"){
        # iwiw sample network
        N <<- 271941
        elist <- fread("R_adj.csv.gz")
        A <<- sparseMatrix(
            c(elist$i+1,elist$j+1),
            c(elist$j+1,elist$i+1),
            dims=c(N,N)
        )
        seed <<- as.vector(fread("R_seed.csv.gz")$V1)+1

    }
    
}

restart <- function(){
# This function restarts the simulation: it sets only the seeds as infected, everyone else susceptible, and resets the time counter to 1.
# Also, it empties the time_infected list that stores the infection time of the nodes.

    # init new simulation
    time_counter <<- 1

    # creating infected and susceptible lists
    infected <<- logical(N)
    # initial infection
    infected[seed] <<- TRUE
    # initial susceptibles, aka everybody who's not infected
    susceptible <<- !infected

    # storing the infection time for the nodes, setting infected nodes times to 1
    time_infected <<- numeric(N)
    time_infected[seed] <<- 1

    # storing number of infected neighbors for nodes
    node_neighborhood_num_infected <<- numeric(N)
    node_neighborhood_num_neighbors <<- rowSums(A)
    # adding 1 for isolated nodes to avoid division by 0
    node_neighborhood_num_neighbors <<- replace(
        node_neighborhood_num_neighbors,
        node_neighborhood_num_neighbors==0,
        1    
    )
    # counting number of infected neighbors for initial infection
    node_neighborhood_num_infected <<- node_neighborhood_num_infected + rowSums(A[,infected])

}

step_time <- function(){
# This is a function that forwards the simulation by 1 timestep.

    # init stopper
    tic()

    # increment time counter
    time_counter <<- time_counter + 1

    # fraction of infected neighbors vs total neighbors
    a <- node_neighborhood_num_infected/node_neighborhood_num_neighbors
    # uniform random number between 0 and 1
    r <- runif(N)
    # calculating threshold, comparing random number to threshold
    mask <- r < p_par + q_par*a
    # infect nodes where random number is smaller than the threshold and that are susceptible
    new_infected <- susceptible & mask

    # store infection time for newly infected nodes
    time_infected[which(new_infected)] <<- time_counter
    # union of old and new infections
    infected <<- infected | new_infected
    # decreasing the susceptibles with the newly infected nodes
    susceptible <<- susceptible & !new_infected
    # incrementing the counters in the array storing the number of infected neighbors
    if (sum(new_infected)==1){
        node_neighborhood_num_infected <<- node_neighborhood_num_infected + rowSums(as.matrix(A[,new_infected]))
    }else{
        node_neighborhood_num_infected <<- node_neighborhood_num_infected + rowSums(A[,new_infected])
    }

    # end stopper
    toc()
    # output time, new infected, overall infected
    cat(paste0(time_counter," ",sum(new_infected)," ",sum(infected)))
    cat("\n")

}


## main code
init()
restart()

if (mode=="iwiw"){
    for (i in 2:stop_par){
    step_time()
    }
}

# result
plot(table(time_infected))
plot(cumsum(table(time_infected)))

